lib/internal/encoding.js

```diff
@@ -0,0 +1,375 @@
+  ['cp1257', 'windows-1257'],
+  ['windows-1257', 'windows-1257'],
+  ['x-cp1257', 'windows-1257'],
+  ['cp1258', 'windows-1258'],
+  ['windows-1258', 'windows-1258'],
+  ['x-cp1258', 'windows-1258'],
+  ['x-mac-cyrillic', 'x-mac-cyrillic'],
+  ['x-mac-ukrainian', 'x-mac-cyrillic'],
+  ['chinese', 'gbk'],
+  ['csgb2312', 'gbk'],
+  ['csiso58gb231280', 'gbk'],
+  ['gb2312', 'gbk'],
+  ['gb_2312', 'gbk'],
+  ['gb_2312-80', 'gbk'],
+  ['gbk', 'gbk'],
+  ['iso-ir-58', 'gbk'],
+  ['x-gbk', 'gbk'],
+  ['gb18030', 'gb18030'],
+  ['big5', 'big5'],
+  ['big5-hkscs', 'big5'],
+  ['cn-big5', 'big5'],
+  ['csbig5', 'big5'],
+  ['x-x-big5', 'big5'],
+  ['cseucpkdfmtjapanese', 'euc-jp'],
+  ['euc-jp', 'euc-jp'],
+  ['x-euc-jp', 'euc-jp'],
+  ['csiso2022jp', 'iso-2022-jp'],
+  ['iso-2022-jp', 'iso-2022-jp'],
+  ['csshiftjis', 'shift_jis'],
+  ['ms932', 'shift_jis'],
+  ['ms_kanji', 'shift_jis'],
+  ['shift-jis', 'shift_jis'],
+  ['shift_jis', 'shift_jis'],
+  ['sjis', 'shift_jis'],
+  ['windows-31j', 'shift_jis'],
+  ['x-sjis', 'shift_jis'],
+  ['cseuckr', 'euc-kr'],
+  ['csksc56011987', 'euc-kr'],
+  ['euc-kr', 'euc-kr'],
+  ['iso-ir-149', 'euc-kr'],
+  ['korean', 'euc-kr'],
+  ['ks_c_5601-1987', 'euc-kr'],
+  ['ks_c_5601-1989', 'euc-kr'],
+  ['ksc5601', 'euc-kr'],
+  ['ksc_5601', 'euc-kr'],
+  ['windows-949', 'euc-kr'],
+  ['utf-16be', 'utf-16be'],
+  ['utf-16le', 'utf-16le'],
+  ['utf-16', 'utf-16le'],
+  // Unsupported standard mappings
+  // ['csiso2022kr', 'replacement'],
+  // ['hz-gb-2312', 'replacement'],
+  // ['iso-2022-cn', 'replacement'],
+  // ['iso-2022-cn-ext', 'replacement'],
+  // ['iso-2022-kr', 'replacement'],
+  // ['x-user-defined', 'x-user-defined'],
+  // ['iso-8859-16', 'iso-8859-16'],
+]);
+
+// Unfortunately, String.prototype.trim also removes non-ascii whitespace,
+// so we have to do this manually
+function trimAsciiWhitespace(label) {
+  var s = 0;
+  var e = label.length;
+  while (s < e && (
+         label[s] === '\u0009' ||
+         label[s] === '\u000a' ||
+         label[s] === '\u000c' ||
+         label[s] === '\u000d' ||
+         label[s] === '\u0020')) {
+    s++;
+  }
+  while (e > s && (
+         label[e - 1] === '\u0009' ||
+         label[e - 1] === '\u000a' ||
+         label[e - 1] === '\u000c' ||
+         label[e - 1] === '\u000d' ||
+         label[e - 1] === '\u0020')) {
+    e--;
+  }
+  return label.slice(s, e);
+}
+
+function getEncodingFromLabel(label) {
+  const enc = encodings.get(label);
+  if (enc !== undefined) return enc;
+  return encodings.get(trimAsciiWhitespace(label.toLowerCase()));
+}
+
+function hasTextDecoder(encoding = 'utf-8') {
+  if (typeof encoding !== 'string')
+    throw new errors.Error('ERR_INVALID_ARG_TYPE', 'encoding', 'string');
+  return hasConverter(getEncodingFromLabel(encoding));
+}
+
+class TextDecoder {
+  constructor(encoding = 'utf-8', options = {}) {
+    if (typeof encoding !== 'string')
+      throw new errors.Error('ERR_INVALID_ARG_TYPE', 'encoding', 'string');
```

Comment:
This should coerce `encoding` to string rather than check its type, per IDL rules. See [here](https://github.com/nodejs/node/blob/448c4c62d2b413226dfdef03d6f8d243de0984a3/lib/internal/url.js#L56) for how it's done in URL.